path_credentials.go

```diff
@@ -39,115 +38,104 @@ func (b *tfBackend) terraformToken() *framework.Secret {
 	if err != nil {
-		return logical.ErrorResponse("Error creating Terraform token: %s", err), err
-	}
-
-	if token == nil {
-		return nil, errors.New("error creating Terraform token")
-	}
-
-	if err := framework.DeleteWAL(ctx, s, walID); err != nil {
-		return nil, errwrap.Wrapf("error deleting WAL: {{err}}", err)
+		return nil, err
 	}
 
 	resp := b.Secret(terraformTokenType).Response(map[string]interface{}{
-		"token": token.Token,
+		"token":    token.Token,
+		"token_id": token.ID,
 	}, map[string]interface{}{
-		"token_id":     token.ID,
-		"organization": roleEntry.Organization,
-		"team_id":      roleEntry.TeamID,
-		"role":         roleName,
-		"description":  token.Description,
+		"token_id": token.ID,
+		"role":     role.Name,
 	})
 
-	if roleEntry.TTL > 0 {
-		resp.Secret.TTL = roleEntry.TTL
+	if role.TTL > 0 {
+		resp.Secret.TTL = role.TTL
 	}
 
-	if roleEntry.MaxTTL > 0 {
-		resp.Secret.MaxTTL = roleEntry.MaxTTL
+	if role.MaxTTL > 0 {
+		resp.Secret.MaxTTL = role.MaxTTL
 	}
 
 	return resp, nil
 }
 
-func (b *tfBackend) credentialRead(ctx context.Context, s logical.Storage, roleName string) (*terraformRoleEntry, error) {
-	if roleName == "" {
-		return nil, errors.New("missing role name")
+func (b *tfBackend) createToken(ctx context.Context, s logical.Storage, roleEntry *terraformRoleEntry) (*terraformToken, error) {
+	client, err := b.getClient(ctx, s)
+	if err != nil {
+		return nil, err
+	}
+
+	var token *terraformToken
+
+	switch {
+	case isOrgToken(roleEntry.Organization, roleEntry.TeamID):
+		token, err = createOrgToken(ctx, client, roleEntry.Organization)
+	case isTeamToken(roleEntry.TeamID):
+		token, err = createTeamToken(ctx, client, roleEntry.TeamID)
+	default:
+		token, err = createUserToken(ctx, client, roleEntry.UserID)
 	}
 
-	entry, err := s.Get(ctx, "roles/"+roleName)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating Terraform token: %w", err)
 	}
 
-	var roleEntry terraformRoleEntry
-	if entry != nil {
-		if err := entry.DecodeJSON(&roleEntry); err != nil {
-			return nil, err
-		}
-		return &roleEntry, nil
+	if token == nil {
+		return nil, errors.New("error creating Terraform token")
 	}
 
-	return nil, nil
+	return token, nil
 }
 
 const pathCredentialsHelpSyn = `
 Generate a Terraform Cloud or Enterprise API token from a specific Vault role.
 `
 
 const pathCredentialsHelpDesc = `
-This path generates Terraform Cloud or Enterprise API Organization or Team
-Token based on a particular role.
+This path generates Terraform Cloud or Enterprise API Organization, Team, or
+User Tokens based on a particular role.
 
-If the role has the organization and team ID configured,
-this path generates a team token.
+If the role has the team ID configured, this path generates a team token.
 
 If this role only has the organization configured, this path generates an
 organization token.
+
+If this role has a user ID configured, this path generates a user token.
```

Comment:
Is it worth mentioning that it will only generate a user token if team ID and organization are not present?

Reply:
Probably so! I updated the description in b38c27a, let me know what you think